src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/TypesTableNode.cs

```diff
@@ -51,7 +51,7 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                     throw new NotImplementedException();
                 }
 
-                int hashCode = type.GetHashCode();
+                int hashCode = GetHashCode((EcmaType)type);
```

Comment:
Type hashing behaves differently in CoreCLR vs CoreRT. In CoreRT we do `Hash(Namespace.TypeName)`. In CoreCLR we `Hash(Namespace) ^ Hash(TypeName)`. We could imagine adding a TypeSystemContext extension point to provide customized type hashing. My feeling is that's overkill right now since the available types table is the only place we need to embed type hashcodes. We have another table for instance method entrypoint that uses method hashcodes, too that might need a similar change.

Reply:
Is there a specific reason why CoreCLR chose to xor these? We can just standardize on whatever produces better distributions. It shouldn't be too hard to compute distributions in a couple assemblies and compare.

In general, doing a simple `xor` of two hashes is not a great way to combine hash codes (not sure in this case). What we do in CoreRT makes it look like we're just hashing a single long namespace. Long namespaces is what we're going to tune the algorithm for, should we need better distributions.

Reply:
CoreCLR's hashing algorithm was implemented that way to avoid a bunch of string concatenation in the runtime since the namespace / type name strings always separate. Array hash codes are also computed differently to avoid string conversion. I'm not sure how valid these perf concerns are vs how much the original implementer just didn't want to deal with native string manipulation. We'd need to analyze from both the hash distribution and the runtime performance angle.

See https://github.com/dotnet/coreclr/blob/024f809869d156f73ea6d7527d9664ce90b5505e/src/vm/typehashingalgorithms.h#L46

Reply:
CoreCLR could just steal the HashCodeBuilder class from CoreRT.

https://github.com/dotnet/corert/blob/1271b47cc5cff1bf10dc4143186fd01f8c76b8f1/src/Common/src/TypeSystem/Common/TypeHashingAlgorithms.cs#L18-L62

This helper struct is how we avoid string concatenation in CoreRT and keep hash code computation allocation free.